util/grpcerrors/grpcerrors.go

```diff
@@ -27,99 +29,306 @@ type TypedErrorProto interface {
+			// Set to unknown just in case it leaks somehow.
+			Code:    int32(codes.Unknown),
+			Message: err.Error(),
 		}
-	})
+		withDetails(detail, err)
 
-	if len(details) > 0 {
-		if st2, err := withDetails(ctx, st, details...); err == nil {
-			st = st2
+		if any, err := anypb.New(detail); err == nil {
+			p.Details = append(p.Details, any)
 		}
 	}
+}
 
-	return st.Err()
+// marshalAny will marshal the error into a typeurl.Any.
+// If this isn't possible, it will marshal it as an empty
+// protobuf to ensure that it is non-nil.
+func marshalAny(err error) (any typeurl.Any) {
+	if typed, ok := err.(TypedError); ok {
+		any = marshalTypedError(typed.ToProto())
+	}
+	if any == nil {
+		any = marshalStacktrace(err)
+	}
+	if any == nil {
+		any, _ = typeurl.MarshalAny(err)
+	}
+	if any == nil {
+		any, _ = anypb.New(&emptypb.Empty{})
+	}
+	return any
 }
 
-func withDetails(ctx context.Context, s *status.Status, details ...proto.Message) (*status.Status, error) {
-	if s.Code() == codes.OK {
-		return nil, errors.New("no error details for status with code OK")
+// marshalTypedError will marshal a TypedError
+// into an Any protobuf with JSON marshaling.
+//
+// If this fails, it will return nil.
+func marshalTypedError(p proto.Message) typeurl.Any {
+	url, err := typeurl.TypeURL(p)
+	if err != nil {
+		return nil
 	}
-	p := s.Proto()
-	for _, detail := range details {
-		url, err := typeurl.TypeURL(detail)
-		if err != nil {
-			bklog.G(ctx).Warnf("ignoring typed error %T: not registered", detail)
-			continue
-		}
-		dt, err := json.Marshal(detail)
-		if err != nil {
-			return nil, err
-		}
-		p.Details = append(p.Details, &any.Any{TypeUrl: url, Value: dt})
+
+	// TODO: This isn't strictly correct, but changing
+	// this at this point would be difficult with existing
+	// serialization and backwards compatibility.
+	//
+	// The typeurl library has a bug in it where a protobuf
+	// message will be serialized as protobuf but it will
+	// attempt to deserialize it with JSON if it is registered
+	// with typeurl.Register. All of our typed errors do that
+	// so all of them encounter this bug. In order to avoid this
+	// bug and maintain backwards compatibility, we force JSON
+	// marshaling and are just really careful about making sure
+	// to register the typeurl.
+	dt, err := json.Marshal(p)
+	if err != nil {
+		return nil
+	}
+
+	return &anypb.Any{
+		TypeUrl: url,
+		Value:   dt,
 	}
-	return status.FromProto(p), nil
+}
+
+func marshalStacktrace(err error) typeurl.Any {
+	var st *stack.Stack
+	switch err := err.(type) {
+	case interface{ StackTrace() pkgerrors.StackTrace }:
+		st = stack.Convert(err.StackTrace())
+	case interface{ StackTrace() *stack.Stack }:
+		st = err.StackTrace()
+	default:
+		return nil
+	}
+
+	// stack.Stack is affected by the same bug as the other
+	// typed errors so reuse that marshaling function to
+	// marshal the stack trace.
+	return marshalTypedError(st)
 }
 
 func Code(err error) codes.Code {
```

Comment:
The code here looks good and favors the right error code but it might be good to define that in the docs.